### DISCUSSION TITLE
Introduce more flexible docker image tagging

### DISCUSSION THREAD

**Discussion by fl0wx** [grafana/grafana#47177] [General]
hey all,

i would propose to do somehing similar to the tagging which go does: https://hub.docker.com/_/golang?tab=description 

so you would have more than one tag referring to a specific image. Currently that would be 

Grafana Image for 8.4.5 has tags:
* 8
* 8.4
* 8.4.5

Grafana Image for 8.3.7 has tags:
* 8.3
* 8.3.7

This would lead to better automatic ugprades on hotfixes / security fixes and more flexible development cycles.

Thanks and kr 
fl0w

**Comment by chennin** [comment]
Today (my first time looking at Grafana in docker) I was surprised to see that "latest" also points to "9.0.0-beta2".  I would also like to see a more "stable" tagging pattern that would let me (1) avoid pinning to a very specific version and (2) not (immediately) get major/beta/breaking versions.  E.g. right now I would want to pin to 8 or 8.5 but not 8.5.4 (note that I do not know if 8.6 or 8.5.5 are planned).

**Reply by eightnoneone** [reply]
Funny enough, I was marching through a series of upgrades to get to 8.5.4 and ended up mistakenly setting `latest` at one point. Now I have a 9.0.0 install that I didn't want. 

Having latest = beta seems a mistake.

**Reply by op3** [reply]
The `latest` tag is just switches wildly back and forth between major versions. My installation just upgraded to 9.0.0-beta3 and then downgraded back to 8.5.5, which of course results in numerous problems because of migrated databases, files, or whatever. Having this tag around is not even useless, it is actively dangerous because any “upgrade” could result in an actual downgrade of Grafana, which will break things.

**Comment by fl0wx** [comment]
I think this can be pretty easy done via adding:
```bash
# Source: https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string & made it POSIX compatible:
semverRegEx='^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)(-((0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*)(\.(0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*))*))?(\+([0-9a-zA-Z-]+(\.[0-9a-zA-Z-]+)*))?$'

if [[ ${_grafana_version} =~ $semverRegEx ]]; then
    docker_tag_all ${BASH_REMATCH[1]} #only for the highest minor version in each major version release, dont know how to implement this.
    docker_tag_all ${BASH_REMATCH[1]}.${BASH_REMATCH[2]} 
fi
```
to https://github.com/grafana/grafana/blob/1d3cd0103ea7a7d9eab900aa6af86b6dadfd035c/packaging/docker/build.sh#L104-L107 or am i missing something?

**Comment by sjwk** [comment]
Definitely agree with the points here.  Being able to lock to a major version would be useful, but absolutely there should be different tags for stable and beta.  I'm following 'latest' which just upgraded to 9.1.0-beta1 which is now giving 'request was aborted' errors for many of my panels.  I'd prefer 'latest' for stable versions, 'beta' or 'test' for development releases, but would be happy with 'stable' and 'latest' for development, as long as it's documented.

**Comment by andonovski** [comment]
Another voice to have separate tag for stable version. It is really need, using tag latest downloads latest beta version, which gives errors in UI like @sjwk already mentioned.
Either latest is only latest stable version, or introduce tag stable. Beta versions are not desirable in production environment.

**Comment by ianw** [comment]
Just to add to this; I filed https://github.com/grafana/grafana/issues/53275 after our production updated to the 9 beta

https://opendev.org grafana instance at https://grafana.opendev.org was running with latest, which pulled the beta and broke.

We have some (if I do say so myself :) pretty clever CI where we keep our db dashboards in configuration [1] and when any changes are proposed, we run a job that starts up a grafana, loads the dashboards and takes screenshots [2] which it saves in the job results.

We definitely like to keep up-to-date and minimise our distance (and hence unknown problems!) from latest releases.  For other projects we consume (Ansible is a good example) we run a parallel CI job in a non-voting (doesn't stop changes merging) mode against the devel/beta/unreleased branch/image/tag/etc. This way we see problems like 53275 early and can start working on (and verifying) fixes, but don't break production.

So for our purposes, having both betas and stable images released in a controlled fashion would be great.

[1] https://opendev.org/openstack/project-config/src/branch/master/grafana
[2] https://zuul.opendev.org/t/openstack/builds?job_name=project-config-grafana&skip=0 - click on a job and see "artifacts -> screenshots"

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
